src/modules/pmc/src/lib/PmcBase.cpp

```diff
@@ -684,29 +752,130 @@ int PmcBase::ConfigureSources(const std::map<std::string, std::string> sources)
-                m_executionState.SetExecutionState(ExecutionState::StateComponent::failed, ExecutionState::SubStateComponent::modifyingSources, source.first);
+                m_executionState.SetExecutionState(StateComponent::Failed, SubStateComponent::ModifyingSources, source.first);
                 output.close();
                 return errno;
             }
 
-            output << source.second << std::endl;
-            output.close();
+            std::string packageSource = source.second;
+
+            if (ValidateAndUpdatePackageSource(packageSource, gpgKeys))
+            {
+                output << packageSource << std::endl;
+                output.close();
+            }
+            else
+            {
+                OsConfigLogError(PmcLog::Get(), "Invalid source format provided for %s. Stopping configuration for further sources", source.first.c_str());
+                m_executionState.SetExecutionState(StateComponent::Failed, SubStateComponent::ModifyingSources, source.first);
+                output.close();
+                return EINVAL;
+            }
         }
     }
 
-    m_executionState.SetExecutionState(ExecutionState::StateComponent::running, ExecutionState::SubStateComponent::updatingPackageLists);
-    status = RunCommand(g_commandAptUpdate, nullptr);
+    m_executionState.SetExecutionState(StateComponent::Running, SubStateComponent::UpdatingPackageLists);
+    status = RunCommand(g_commandAptUpdate, nullptr, true);
 
-    if (status != 0)
+    if (status != PMC_0K)
     {
         OsConfigLogError(PmcLog::Get(), "Refresh package lists failed with status %d", status);
-        status == ETIME ? m_executionState.SetExecutionState(ExecutionState::StateComponent::timedOut, ExecutionState::SubStateComponent::updatingPackageLists)
-            : m_executionState.SetExecutionState(ExecutionState::StateComponent::failed, ExecutionState::SubStateComponent::updatingPackageLists);
+        status == ETIME ? m_executionState.SetExecutionState(StateComponent::TimedOut, SubStateComponent::UpdatingPackageLists)
+            : m_executionState.SetExecutionState(StateComponent::Failed, SubStateComponent::UpdatingPackageLists);
     }
     else
     {
-        m_executionState.SetExecutionState(ExecutionState::StateComponent::succeeded, ExecutionState::SubStateComponent::none);
+        m_executionState.SetExecutionState(StateComponent::Succeeded, SubStateComponent::None);
+    }
+
+    return status;
+}
+
+bool PmcBase::ValidateAndUpdatePackageSource(std::string& packageSource, const std::map<std::string, std::string>& gpgKeys)
+{
+    std::smatch sourceMatches;
+
+    if (std::regex_match(packageSource, sourceMatches, std::regex(g_regexSources)))
+    {
+        std::smatch signedByMatches;
+        std::string matchedString = sourceMatches[2].matched ? sourceMatches[2].str() : std::string();
```

Comment:
Are these indices always valid on partial matches? Would it be safer to validate `sourceMatches.size() >= 2`?

Reply:
the array size is tightly coupled to the number of capture groups present in the regex ([details here](https://www.geeksforgeeks.org/smatch-regex-regular-expressions-in-c/)).

But since you're the second reviewer with the same concerns 😉, I'll add an explicit check.

Reply:
done!